src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UnconfiguredProjectTasksService.cs

```diff
@@ -3,20 +3,42 @@
 using System;
 using System.ComponentModel.Composition;
 using System.Threading.Tasks;
-
 using Microsoft.VisualStudio.Threading;
 
 namespace Microsoft.VisualStudio.ProjectSystem
 {
+    [Export]
     [Export(typeof(IUnconfiguredProjectTasksService))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
     internal class UnconfiguredProjectTasksService : IUnconfiguredProjectTasksService
     {
         private readonly IProjectAsynchronousTasksService _tasksService;
+        private readonly ILoadedInHostListener _loadedInHostListener;
+        private readonly TaskCompletionSource<object> _projectLoadedInHost = new TaskCompletionSource<object>();
+        private readonly TaskCompletionSource<object> _prioritizedProjectLoadedInHost = new TaskCompletionSource<object>();
 
         [ImportingConstructor]
-        public UnconfiguredProjectTasksService([Import(ExportContractNames.Scopes.UnconfiguredProject)]IProjectAsynchronousTasksService tasksService)
+        public UnconfiguredProjectTasksService([Import(ExportContractNames.Scopes.UnconfiguredProject)]IProjectAsynchronousTasksService tasksService, ILoadedInHostListener loadedInHostListener)
         {
             _tasksService = tasksService;
+            _loadedInHostListener = loadedInHostListener;
+        }
+
+        [ProjectAutoLoad(completeBy:ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
+        public Task OnProjectFactoryCompleted()
+        {
+            return _loadedInHostListener.StartListeningAsync();
+        }
+
+        public Task ProjectLoadedInHost
+        {
+            get { return _projectLoadedInHost.Task.WithCancellation(_tasksService.UnloadCancellationToken); }
```

Comment:
Undecided whether to do this - *or* CancellationToken.Register. Was worried that the latter would get out of hand for large solutions as you can never unregister.